src/docker-client-factory.ts

```diff
@@ -1,22 +1,29 @@
 import Dockerode from "dockerode";
+import { Writable } from "stream";
 import url from "url";
 import { DockerClient, DockerodeClient } from "./docker-client";
+import { FileHelper, RealFileHelper } from "./file-helper";
 import log from "./logger";
+import ProcessEnv = NodeJS.ProcessEnv;
 
 export type Host = string;
 
 export interface DockerClientFactory {
-  getClient(): DockerClient;
-  getHost(): Host;
+  getClient(): Promise<DockerClient>;
+  getHost(): Promise<Host>;
 }
 
 export class DockerodeClientFactory implements DockerClientFactory {
-  private readonly host: Host;
-  private readonly client: DockerClient;
+  private readonly host: Promise<Host>;
+  private readonly client: Promise<DockerClient>;
 
-  constructor() {
-    if (process.env.DOCKER_HOST) {
-      const { host, client } = this.fromDockerHost(process.env.DOCKER_HOST);
+  constructor(env: ProcessEnv = process.env, private readonly fileHelper: FileHelper = new RealFileHelper()) {
+    if (env.DOCKER_HOST) {
```

Comment:
What if someone sets `DOCKER_HOST=unix:///var/run/docker.sock`?

Reply:
Is there a use-case for it?

Reply:
If someone configures their GitlabCI docker runner with `DOCKER_HOST=unix:///var/run/docker.sock` then it should still work, shouldn't it (despite this configuration being useless because it is the default)?

Imo, the distinction should not be done based on the presence of a environment variable, but rather based on its content (which includes the default if it is not present).